### PR TITLE
Assorted kqueue fixes/improvements

### PR DESCRIPTION
**Pull request by nagisa**
See the commits for description

### CODE REVIEW

File: src/sys/unix/kqueue.rs [hunk 1]
```diff
@@ -35,7 +36,7 @@ impl Selector {
         Ok(Selector {
             id: id,
             kq: kq,
-            changes: RefCell::new(KeventList(Vec::new())),
+            changes: RefCell::new(KeventList(Vec::with_capacity(32))),
```
Comment (nagisa) [inline src/sys/unix/kqueue.rs]: I feel like a better option here would be to use a statically sized ring-like buffer, but that would increase size of Selector significantly, which I’m not sure is desired.

**Comment by carllerche** [comment]
Thanks for the PR.

The commit that removes the RefCell looks fine to me.

I do have a question about removing the `EV_DISABLE`. The reason it was originally used was to handle `reregister`. Specifically, if you register a socket w/ `Readable + Writable`, then you reregister it with only `Readable`, that should prevent any writable events from being fired. Since kqueue tracks readable and writable separately, it seems that this change would mean that there is no way to "reregister" a FD to remove an interest. Is this correct?

**Comment by nagisa** [comment]
> I do have a question about removing the EV_DISABLE. The reason it was originally used was to handle reregister. Specifically, if you register a socket w/ Readable + Writable, then you reregister it with only Readable, that should prevent any writable events from being fired. Since kqueue tracks readable and writable separately, it seems that this change would mean that there is no way to "reregister" a FD to remove an interest. Is this correct?

In the 2nd commit I take care to ensure mio `EV_DELETE`s the writable event in the scenario you describe, essentially preserving the semantics of the code previously.

Since in the 2nd commit code also checks registration errors on a per-`kevent` basis,  so it is able to precisely ignore the errors that arise when event that’s being `EV_DELETE`d does not exist. This is still preferable over the `EV_DISABLE` approach, due to problem described in the 1st commit (not all kinds of descriptors can be registered with `EV_WRITE|EV_DISABLE`).

`EV_DELETE` approach may also be more efficient during actual polling, but I haven’t confirmed that.

**Comment by carllerche** [comment]
Ok, that makes sense. I think this is a good change.

**Comment by carllerche** [comment]
Hmm, so the tests don't pass locally for me (OS X 10.12.2)

It looks like registering a kqueue handle w/ write interest works?

```

running 1 test
test sys::unix::kqueue::does_not_register_rw ... FAILED

failures:

---- sys::unix::kqueue::does_not_register_rw stdout ----
	thread 'sys::unix::kqueue::does_not_register_rw' panicked at 'called `Result::unwrap_err()` on an `Ok` value: ()', ../src/libcore/result.rs:837
note: Run with `RUST_BACKTRACE=1` for a backtrace.

**Comment by nagisa** [comment]
It passes on Travis’ 10.11 OS X and all of the BSDs I’ve tested (which is FreeBSD). Must be a OS X bug or maybe undocumented (apple likes those) and intended change.

I’m fine with removing the failing check, as really only the other one matters.

**Comment by carllerche** [comment]
Thanks, merged 11e6cb32df2152cd2b0fd9668f42547f55bcdbc5

**Pull request closed without merging**